local-cluster/tests/local_cluster.rs

```diff
@@ -4480,6 +4480,216 @@ fn test_slot_hash_expiry() {
+    }
+
+    // Kill A and restart B with voting. B should now fork off
+    info!("Killing A");
+    let a_info = cluster.exit_node(&a_pubkey);
+
+    info!("Restarting B");
+    b_info.config.voting_disabled = false;
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    // Allow B to fork now.
+    info!("Allowing B to fork");
+    let now = Instant::now();
+    loop {
+        sleep(Duration::from_millis(200));
+        let (last_vote, _) = last_vote_in_tower(&b_ledger_path, &b_pubkey).unwrap();
+        if last_vote > common_ancestor_slot + 3 {
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(10),
+            "validator not creating blocks fast enough"
+        );
+    }
+    let last_vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+    let blockstore = open_blockstore(&b_ledger_path);
+    info!(
+        "B last vote: {} minority fork: {:?}",
+        last_vote_on_b,
+        AncestorIterator::new(last_vote_on_b, &blockstore).collect::<Vec<Slot>>()
+    );
+
+    info!("Kill B");
+    b_info = cluster.exit_node(&b_pubkey);
+
+    info!("Resolve the partition");
+    {
+        // Here we let B know about the missing blocks that A had produced on its partition
+        let a_blockstore = open_blockstore(&a_ledger_path);
+        let b_blockstore = open_blockstore(&b_ledger_path);
+        copy_blocks(last_vote_on_a, &a_blockstore, &b_blockstore);
+    }
+
+    // Now restart A and B and see if B is able to eventually send a new vote on tip of its fork.
+    info!("Restarting A & B");
+    cluster.restart_node(&a_pubkey, a_info, SocketAddrSpace::Unspecified);
+    cluster.restart_node(&b_pubkey, b_info, SocketAddrSpace::Unspecified);
+
+    let vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+    assert!(
+        vote_on_b == last_vote_on_b,
+        "B should not have new votes immediately"
+    );
+
+    let blockstore = open_blockstore(&a_ledger_path);
+    info!(
+        "A majority fork: {:?}",
+        AncestorIterator::new(last_vote_on_a, &blockstore).collect::<Vec<Slot>>()
+    );
+
+    let now = Instant::now();
+    loop {
+        sleep(Duration::from_millis(500));
+        let vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+        if vote_on_b > last_vote_on_b {
+            info!(
+                "B has voted again: {} last vote {} elapsed time {:?}",
+                vote_on_b,
+                last_vote_on_b,
+                now.elapsed(),
+            );
+            break;
+        }
+        assert!(
+            now.elapsed() < Duration::from_secs(60),
+            "B hasn't voted after a long time"
+        );
+    }
+    let vote_on_b = wait_for_last_vote_in_tower_to_land_in_ledger(&b_ledger_path, &b_pubkey);
+    // check that B has voted past last voted slot.
+    assert!(vote_on_b > last_vote_on_b);
```

Comment:
1.  one more good check would be to check that the tower has no additional votes between `last_vote_on_b` and `vote_on_b`, to make sure that this is the first additional vote. 
2. would be good to check that `last_vote_on_b - vote_on_b` the distance is greater than `MAX_SLOT_HASHES`

Reply:
Done.